### PR TITLE
add minibatching

### PR DESCRIPTION
**Pull request by lvwerra**
Until now the PPO mini batch size has been hardcoded to 1. This PR aims to change it by refactoring the forward/backward passing logic.

**In summary this PR does the following things:**
1. The `batched_forward_pass` returns new a `mask` which can be used to mask parts of the sequence to be ignored
2. enable mini-batching of PPO by creating a small `dataloader` with the `mini_batch_size` to sample from the current PPO batch
3. In the `loss` method we replace all operations affected by masked parts of the sequence with masked ones (`masked_mean`, `masked_whiten`)
4. remove `compute_logits_vpred ` and use `batched_forward_pass` for everything
5. extend testing and refactor it (i don't think we need subfolders for the 3 test files we have)

**W&B logs:**
- [GPT-2 (mini_batch_size=1)](https://wandb.ai/lvwerra/trl/runs/szaov4b4)
- [GPT-2 (mini_batch_size=4)](https://wandb.ai/lvwerra/trl/runs/2knwk29s)
- [T5 (mini_batch_size=1)](https://wandb.ai/lvwerra/trl/runs/uz003w2f)
- [T5 (mini_batch_size=4)](https://wandb.ai/lvwerra/trl/runs/5j24hepl)

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this great addition! I left few comments and questions as a first pass!

**Review by younesbelkada — COMMENTED** [review]
Also, what about completely removing `forward_batch_size` from the config? I don't think this is a breaking change as the configs cannot be pushed on the Hub, just need to update the examples accodingly. I believe this can be done on a follow up PR too

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -599,20 +639,30 @@ def batched_forward_pass(
...
+            all_values.append(values)
+            all_logprobs.append(logprobs)
+            all_masks.append(masks)
+
+        return (
+            torch.cat(all_logprobs),
+            torch.cat(all_logits)[:, :-1],
+            torch.cat(all_values)[:, :-1],
+            torch.cat(all_masks)[:, :-1],
+        )
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:656]: Since start/end is potentially different for each sample we wouldn't be able to stack them. The mask takes care of "slicing" by "masking" the tensor entries that are not used. the `[:-1]` is because we get predictions for one more token than we need. alternatively we could avoid this by removing the last `input_id` from the forward pass. No strong opinion :)

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -538,42 +601,18 @@ def batched_forward_pass(
                     shape (`batch_size`, `response_length`)
                 - all_values (`torch.FloatTensor`): Values of the responses, shape (`batch_size`, `response_length`)
         """
-        bs = self.config.batch_size
-        fbs = self.config.forward_batch_size
+        bs = len(queries)
+        fbs = min(bs, self.config.forward_batch_size)
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:605]: It's for the case where `mini_batch_size` is smaller than `forward_batch_size` during the forward passes inside the minibatch loop. I am also not quite happy with how we do it actually.

**Comment by lvwerra** [comment]
The breaking change actually also happens for users who currently use the library with `forward_batch_size`. What do you think about setting it default to `None` and overwrite `mini_batch_size` if it's set to another value with a warning that it affects now also the  `mini_batch_size` if set to a value?

**Comment by younesbelkada** [comment]
This solution makes a lot of sense yes!

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -798,12 +830,17 @@ def record_step_stats(self, kl_coef: float, **data):
...
-            torch.stack([torch.sum(non_score_reward) for non_score_reward in data["non_score_reward"]])
-        )
+        mask = data.pop("masks")
+
+        kl_list = ((data["logprobs"] - data["ref_logprobs"]) * mask).sum(axis=-1)
+        mean_kl = kl_list.mean()
+        mean_entropy = (-data["logprobs"] * mask).sum(axis=-1).mean()
+        # kl_list = [logprobs - ref_logprobs for logprobs, ref_logprobs in zip(data["logprobs"], data["ref_logprobs"])]
+        # mean_kl = masked_mean(torch.stack([torch.sum(kl) for kl in kl_list]), mask)
+        # mean_entropy = masked_mean(torch.stack([torch.sum(-log_probs) for log_probs in data["logprobs"]]), mask)
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:840]: good catch :)

**Comment by lvwerra** [comment]
Deprecated `forward_batch_size`: feel free to have a look!

File: trl/trainer/ppo_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:648]: I see this makes sense!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your great work on this! 💯
